packages/qwik/src/core/v2/ssr/vnode-data.unit.tsx

```diff
@@ -0,0 +1,215 @@
+import { describe, expect, it } from 'vitest';
+import { componentQrl } from '../../component/component.public';
+import { inlinedQrl } from '../../qrl/qrl';
+import { useSignal } from '../../use/use-signal';
+import { ssrRenderToDom } from '../rendering.unit-util';
+import { codeToName } from '../shared/shared-serialization';
+import { encodeAsAlphanumeric } from './vnode-data';
+
+const debug = false;
+
+describe('vnode data', () => {
+  describe('encodeAsAlphanumeric', () => {
+    it('should return A for 0', () => {
+      expect(encodeAsAlphanumeric(0)).toEqual('A');
+    });
+    it('should return B for 1', () => {
+      expect(encodeAsAlphanumeric(1)).toEqual('B');
+    });
+    it('should return K for 10', () => {
+      expect(encodeAsAlphanumeric(10)).toEqual('K');
+    });
+    it('should return Z for 25', () => {
+      expect(encodeAsAlphanumeric(25)).toEqual('Z');
+    });
+    it('should return bA for 26', () => {
+      expect(encodeAsAlphanumeric(26)).toEqual('bA');
+    });
+    it('should return dW for 100', () => {
+      expect(encodeAsAlphanumeric(100)).toEqual('dW');
+    });
+    it('should return bmM for 1000', () => {
+      expect(encodeAsAlphanumeric(1000)).toEqual('bmM');
+    });
+    it('should return ouQ for 10000', () => {
+      expect(encodeAsAlphanumeric(10000)).toEqual('ouQ');
+    });
+  });
+  describe('integration tests', () => {
+    it('components inside the div', async () => {
+      const Component1 = componentQrl(
+        inlinedQrl(() => {
+          const data = useSignal(1);
+          return (
+            <div>
+              <span>{data.value}</span>
+            </div>
+          );
+        }, 's_cmp1')
+      );
+      const Parent = componentQrl(
+        inlinedQrl(() => {
+          return (
+            <div>
+              <Component1 />
+              <Component1 />
+            </div>
+          );
+        }, 's_parent')
+      );
+
+      const { container } = await ssrRenderToDom(<Parent />, { debug });
+      const state = container.$rawStateData$;
+      const vnodeData = getVNodeDataFromQwikState(state);
+
+      expect(vnodeData).toEqual(['3A', '4A', '4B']);
```

Comment:
I don't have an answer for you, but I would like to figure out something better here.

The issue is that:
- You have an integration test, but really, it asserts an implementation detail. If the implementation details change, it will be hard to know what the test should assert.
- Ideally this integration tests, should be integration all the way through. Could you assert that a correct VNode can be retrieved?

I know I don't have answers here, but was hoping that maybe you could come up with something.